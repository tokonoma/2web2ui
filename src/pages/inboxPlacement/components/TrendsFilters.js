import React, { useCallback, useEffect, useMemo } from 'react';
import moment from 'moment';
import { Tag } from 'src/components/matchbox';
import InboxPlacementTypeahead from 'src/components/typeahead/InboxPlacementTypeahead';
import DatePicker from 'src/components/datePicker/DatePicker';
import { InlineItems } from 'src/components/structure';
import {
  RELATIVE_DATE_OPTIONS,
  RELATIVE_DATE_RANGE_OPTIONS_MAP,
  TAG_LABELS,
} from '../constants/filters';
import { getRelativeDates, formatApiTimestamp } from 'src/helpers/date';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import OGStyles from './TrendsFilters.module.scss';
import HibanaStyles from './TrendsFiltersHibana.module.scss';

const TrendsFilters = ({ filters = {}, updateFilters, validateDate }) => {
  const styles = useHibanaOverride(OGStyles, HibanaStyles);
  const { dateRange: { from, to } = {}, range, tags } = filters;

  // If the "to" is the current time, use the correct relative range option
  // Otherwise, we're using a custom range
  useEffect(() => {
    if (moment.utc().diff(moment.utc(to), 'minute') > 0) {
      updateFilters({
        range: 'custom',
      });
    } else {
      updateFilters({
        range: RELATIVE_DATE_RANGE_OPTIONS_MAP[moment.utc(to).diff(moment.utc(from), 'days')],
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const filterTags = useMemo(
    () =>
      Object.keys(tags).reduce((acc, curr) => {
        const vals = tags[curr].split(',');

        acc = acc.concat(
          vals.map(filter => ({
            type: curr,
            value: filter,
          })),
        );
        return acc;
      }, []),
    [tags],
  );

  const handleFilterRemove = useCallback(
    ({ type, value }) => {
      updateFilters({
        tags: {
          ...tags,
          [type]: tags[type]
            .split(',')
            .filter(val => val !== value)
            .join(','),
        },
      });
    },
    [tags, updateFilters],
  );

  const handleFilterAdd = (item, downshift = {}) => {
    if (item) {
      const currentFilter = tags[item.type];
      downshift.clearSelection();
      updateFilters({
        tags: {
          ...tags,
          [item.type]: currentFilter ? [tags[item.type], item.value].join(',') : item.value,
        },
      });
    }
  };

  return (
    <>
      <div className={styles.Filters}>
        <DatePicker
          relativeDateOptions={RELATIVE_DATE_OPTIONS}
          relativeRange={range}
          from={new Date(from)}
          to={new Date(to)}
          disabled={false}
          validate={validateDate}
          onChange={({ from, to, relativeRange }) => {
            if (from === undefined && to === undefined) {
              const relativeDates = getRelativeDates(relativeRange);

              updateFilters({
                dateRange: {
                  from: formatApiTimestamp(relativeDates.from),
                  to: formatApiTimestamp(relativeDates.to),
                },
                range: relativeRange,
              });
            } else {
              updateFilters({
                dateRange: {
                  from: formatApiTimestamp(from),
                  to: formatApiTimestamp(to),
                },
                range: 'custom',
              });
            }
          }}
          roundToPrecision={true}
        />
        <InboxPlacementTypeahead filters={filters} onChange={handleFilterAdd} />
      </div>
      {Boolean(filterTags.length) && (
        <div className={styles.SelectedFilters}>
          <div className={styles.SelectedFiltersLabel}>Filters:</div>
          <InlineItems>
            {filterTags.map(({ type, value }, index) => (
              <Tag
                key={`${type}_${value}_${index}`}
                onRemove={() => handleFilterRemove({ type, value })}
              >
                {TAG_LABELS[type]}: {value}
              </Tag>
            ))}
          </InlineItems>
        </div>
      )}
    </>
  );
};

export default TrendsFilters;
