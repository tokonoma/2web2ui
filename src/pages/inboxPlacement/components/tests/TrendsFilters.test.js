import { shallow } from 'enzyme';
import moment from 'moment';
import React from 'react';
import TrendsFilters from '../TrendsFilters';

jest.mock('src/hooks/useHibanaOverride', () => styles => styles);

const now = moment.utc(new Date('2019-08-10T12:30:00-04:00'));
const from = moment(now).subtract(30, 'd');
const to = now;

// TODO: Convert this to a mounted test so we can test the
// useEffect that updates the filter range to custom if the
// now is out of date

describe('Inbox Placement Trends Filters', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      updateFilters: jest.fn(),
      validateDate: jest.fn(),
      filters: {
        dateRange: {
          from,
          to,
        },
        tags: {},
      },
    };

    return shallow(<TrendsFilters {...defaultProps} {...props} />);
  };

  it('Matches snapshot test with default filters available', () => {
    const wrapper = subject();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper).not.toHaveTextContent('Filters:');
  });

  it('Displays tags in the filters', () => {
    const filters = {
      dateRange: {
        from,
        to,
      },
      tags: {
        from_domains: 'test.com',
      },
    };
    const wrapper = subject({ filters });
    expect(wrapper).toHaveTextContent('Filters:');
  });
});
