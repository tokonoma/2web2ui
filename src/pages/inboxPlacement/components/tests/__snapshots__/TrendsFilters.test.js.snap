// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbox Placement Trends Filters Matches snapshot test with default filters available 1`] = `
<Fragment>
  <div
    className="Filters"
  >
    <DatePicker
      disabled={false}
      from={2019-07-11T16:30:00.000Z}
      id="date-picker"
      onChange={[Function]}
      preventFuture={true}
      relativeDateOptions={
        Array [
          "7days",
          "30days",
          "90days",
          "custom",
        ]
      }
      roundToPrecision={true}
      to={2019-08-10T16:30:00.000Z}
      validate={[MockFunction]}
    />
    <Connect(InboxPlacementTypeahead)
      filters={
        Object {
          "dateRange": Object {
            "from": "2019-07-11T16:30:00.000Z",
            "to": "2019-08-10T16:30:00.000Z",
          },
          "tags": Object {},
        }
      }
      onChange={[Function]}
    />
  </div>
</Fragment>
`;
